src/ILLink.CodeFix/DAMCodeFixProvider.cs

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using ILLink.CodeFixProvider;
+using ILLink.RoslynAnalyzer;
+using ILLink.Shared;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Simplification;
+
+namespace ILLink.CodeFix
+{
+	public class DAMCodeFixProvider : Microsoft.CodeAnalysis.CodeFixes.CodeFixProvider
+	{
+		public static ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => DynamicallyAccessedMembersAnalyzer.GetSupportedDiagnostics ();
+
+		public sealed override ImmutableArray<string> FixableDiagnosticIds => SupportedDiagnostics.Select (dd => dd.Id).ToImmutableArray ();
+
+		private protected static LocalizableString CodeFixTitle => new LocalizableResourceString (nameof (Resources.DynamicallyAccessedMembersCodeFixTitle), Resources.ResourceManager, typeof (Resources));
+
+		private protected static string FullyQualifiedAttributeName => DynamicallyAccessedMembersAnalyzer.FullyQualifiedDynamicallyAccessedMembersAttribute;
+
+		public sealed override Task RegisterCodeFixesAsync (CodeFixContext context) => BaseRegisterCodeFixesAsync (context);
+
+		static ImmutableHashSet<string> AttriubuteOnReturn ()
+		{
+			var diagDescriptorsArrayBuilder = ImmutableHashSet.CreateBuilder<string> ();
+			diagDescriptorsArrayBuilder.Add (DiagnosticId.DynamicallyAccessedMembersOnMethodReturnValueCanOnlyApplyToTypesOrStrings.AsString ());
+			diagDescriptorsArrayBuilder.Add (DiagnosticId.DynamicallyAccessedMembersMismatchOnMethodReturnValueBetweenOverrides.AsString ());
+			return diagDescriptorsArrayBuilder.ToImmutable ();
+		}
+
+		protected static SyntaxNode[] GetAttributeArguments (ISymbol attributableSymbol, ISymbol targetSymbol, SyntaxGenerator syntaxGenerator, Diagnostic diagnostic)
+		{
+			var symbolDisplayName = targetSymbol.GetDisplayName ();
+			if (string.IsNullOrEmpty (symbolDisplayName) || HasPublicAccessibility (attributableSymbol))
```

Comment:
`IsNullOrEmpty` is usually wrong -- the source should probably pick either null or empty to return if it wants a sentinel value.

Also, the call to `HasPublicAccessibility` looks wrong here. I think that makes sense for RUC but not for DAM.

Reply:
The `GetDisplayName` function looks like it should never return null, so maybe we just check for empty?

`HasPublicAccessibility` was left over from `BaseAttributeCodeFixProvider`, and it's been used in each of the code fixers - let me know if it is relevant, otherwise I'll just remove it

Reply:
Yeah I would just remove anything that was from the `Base` that doesn't make sense here.

Reply:
Also, is there any place that GetDisplayName can return empty string?

We should either have a test case for that, or remove it if it's impossible

Reply:
It was pointed out to me that symbolDisplayName is unused, so we don't need this check at all. I've taken it out